deepchem/models/jax_models/pinns_model.py

```diff
@@ -0,0 +1,391 @@
+'''
+This python consists of diffrent variations of PINNs model using the JaxModel API
+'''
+import numpy as np
+import time
+from typing import Any, Callable, Iterable, List, Optional, Tuple, Union
+
+try:
+  from collections.abc import Sequence as SequenceCollection
+except:
+  from collections import Sequence as SequenceCollection
+
+from deepchem.data import Dataset
+from deepchem.models.losses import Loss
+from deepchem.models.optimizers import Optimizer
+from deepchem.utils.typing import LossFn, OneOrMany
+from deepchem.trans.transformers import Transformer, undo_transforms
+
+# JAX dependencies
+import jax
+import jax.numpy as jnp
+import haiku as hk
+import optax
+from deepchem.models.jax_models.jax_model import JaxModel
+from deepchem.models.jax_models.jax_model import create_default_gradient_fn, create_default_eval_fn
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def create_default_update_fn(optimizer, model_loss):
+  """
+  This function calls the update function, to implement the backpropogation
+  """
+
+  @jax.jit
+  def update(params, opt_state, batch, target, weights,
+             rng) -> Tuple[hk.Params, optax.OptState, jnp.ndarray]:
+    batch_loss, grads = jax.value_and_grad(model_loss)(params, target, weights,
+                                                       rng, *batch)
+    updates, opt_state = optimizer.update(grads, opt_state)
+    new_params = optax.apply_updates(params, updates)
+    return new_params, opt_state, batch_loss
+
+  return update
+
+
+class PinnModel(JaxModel):
+  """
+  This is class is derived from the JaxModel class and methods are also very similar to JaxModel,
+  but it has the optional of passing multiple arguments(Done using *args) suitable for PINNs model.
+  Ex - Approximating f(x, y, z, t) satisfying a Linear diffrential equation.
+
+  This model is reccomended for Linear diffrential equations but if you can accurately write
+  the gradient function in Jax depending on your use case, then it will work as well.
+
+  This class requires two functions apart from the usual function definition and weights
+
+  [1] **grad_fn** : Each PINNs have a diffrent stratergy for calculating its final losses. This 
+  function tells the PinnModel on how to go about computing the derivatives for backpropogation.
+  It should follow this format:
+
+  >>>
+  >> def gradient_fn(forward_fn, loss_outputs, initial_data):
```

Comment:
This is quite a lot to cover here, I wonder if it may be more useful to succinctly list the unique components needed for a PINN model, and then put a minimal working example in the Examples docstring. You can add a link to the tutorial nb later on as well.